### PR TITLE
Refactor dataloading

### PR DESCRIPTION
**Pull request by classicsong**
*Issue #, if available:*
#755 
#756 

*Description of changes:*
This PR include the code refactoring of GraphStorm ``DataLoader`` and ``GraphStorm`` Dataset. All related codes, examples and docs are updated accordingly.

The key changes are:
1. Unify the implementation of different ``GSgnnData`` sub classes (``GSgnnNodeTrainData``, ``GSgnnNodeInferData``, ``GSgnnEdgeTrainData``,  ``GSgnnEdgeInferData``) into ``GSgnnData``.  ``GSgnnData`` now only provides interfaces for accessing graph data, e.g., node features, edge features, labels, train masks, etc.
2. Changes of ``GSgnnData`` API
  * Update the init arguments of ``GSgnnData`` from ``(graph_name, part_config, node_feat_field, edge_feat_field, decoder_edge_feat, lm_feat_ntypes, lm_feat_etypes)`` to ``(part_config, node_feat_field, edge_feat_field, lm_feat_ntypes, lm_feat_etypes)``.
  * Add property functions:
    * ``graph_name()``, return a string of the graph name value in config json.
  * Add new functions:
    * ``get_node_feats(self, input_nodes, nfeat_fields, device='cpu')``, given the node ids (``input_nodes``) and node features to retrieve from the graph data (``nfeat_fields``), return the corresponding node features.
     * ``get_edge_feats(self, input_edges, efeat_fields, device='cpu')``, given the edge ids (``input_edges``) and edge features to retrieve from the graph data (``efeat_fields``), return the corresponding edge features.
     * ``get_node_train_set(self, ntypes, mask)``, return the node training set.
     * ``get_node_val_set(self, ntypes, mask)``, return the node validation set.
     * ``get_node_test_set(self, ntypes, mask)``, return the node test set.
     * ``get_node_infer_set(self, ntypes, mask)``, return the node inference set.
     *  ``get_edge_train_set(self, etypes, mask, reverse_edge_types_map)``, return the edge training set.
     * ``get_node_val_set(self, etypes, mask, reverse_edge_types_map)``, return the edge validation set.
     * ``get_node_test_set(self, etypes, mask, reverse_edge_types_map)``, return the edge test set.
     * ``get_node_infer_set(self, etypes, mask, reverse_edge_types_map)``, return the edge inference set.
  * Update some functions:
    * ``get_node_feats(self, input_nodes, nfeat_fields, device='cpu')``, requires the caller to provide the node feature fields to access
    * ``get_edge_feats(self, input_edges, efeat_fields, device='cpu')``, requires the caller to provide the edge feature fields to access
    * ``get_labels`` function is replaced with ``get_node_feats`` and ``get_edge_feats`` with label field names.
3. Change of GSgnn Dataloaders API:
  * ``GSgnnNodeDataLoaderBase`` and its subclasses, requires three new init arguments:
    * ``label_field``: Label field of the node task.
    * ``node_feats``: Node feature fields used by the node task.
    * ``edge_feats``: Edge feature fields used by the node task.
  * ``GSgnnEdgeDataLoader`` and its subclasses, requires four new init arguments:
    * ``label_field``: Label field of the edge task.
    * ``node_feats``: Node feature fields used by the edge task.
    * ``edge_feats``: Edge feature fields used by the edge task.
    * ``decoder_edge_feats``: Edge feature fields used in the edge task decoder.
  * ``GSgnnLinkPredictionDataLoader `` and its subclasses, requires three new init arguments:
    * ``node_feats``: Node feature fields used by the link prediction task.
    * ``edge_feats``: Edge feature fields used by the link prediction task.
    * ``pos_graph_edge_feats``: The field of the edge features used by positive graph in link prediction.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Review by zhjwy9343 — APPROVED** [review]

File: python/graphstorm/run/gsgnn_ep/ep_infer_lm.py [hunk 1]
```diff
@@ -46,26 +46,32 @@ def main(config_args):
...
     gs.initialize(ip_config=config.ip_config, backend=config.backend,
                   local_rank=config.local_rank)
-
-    infer_data = GSgnnEdgeInferData(config.graph_name,
-                                    config.part_config,
-                                    eval_etypes=config.target_etype,
-                                    node_feat_field=config.node_feat_name,
-                                    label_field=config.label_field,
-                                    decoder_edge_feat=config.decoder_edge_feat)
+    # language model only encoder does not allow node and edge features
```
Comment (zhjwy9343) [inline python/graphstorm/run/gsgnn_ep/ep_infer_lm.py:49]: Not understand what this comment wants to explain.

File: python/graphstorm/run/gsgnn_ep/gsgnn_lm_ep.py [hunk 2]
```diff
@@ -60,12 +60,9 @@ def main(config_args):
                   local_rank=config.local_rank)
     rt_profiler.init(config.profile_path, rank=gs.get_rank())
     sys_tracker.init(config.verbose, rank=gs.get_rank())
-    train_data = GSgnnEdgeTrainData(config.graph_name,
-                                    config.part_config,
-                                    train_etypes=config.target_etype,
-                                    node_feat_field=config.node_feat_name,
-                                    label_field=config.label_field,
-                                    decoder_edge_feat=config.decoder_edge_feat)
+    # language model only encoder does not allow node and edge features
```
Comment (zhjwy9343) [inline python/graphstorm/run/gsgnn_ep/gsgnn_lm_ep.py:63]: ditto

File: python/graphstorm/run/gsgnn_lp/gsgnn_lm_lp.py [hunk 3]
```diff
@@ -79,12 +79,9 @@ def main(config_args):
                   local_rank=config.local_rank)
     rt_profiler.init(config.profile_path, rank=gs.get_rank())
     sys_tracker.init(config.verbose, rank=gs.get_rank())
-    train_data = GSgnnLPTrainData(config.graph_name,
-                                  config.part_config,
-                                  train_etypes=config.train_etype,
-                                  eval_etypes=config.eval_etype,
-                                  node_feat_field=config.node_feat_name,
-                                  pos_graph_feat_field=config.lp_edge_weight_for_loss)
+    # language model only encoder does not allow node and edge features
```
Comment (zhjwy9343) [inline python/graphstorm/run/gsgnn_lp/gsgnn_lm_lp.py:82]: ditto

File: python/graphstorm/run/gsgnn_lp/lp_infer_lm.py [hunk 4]
```diff
@@ -38,30 +38,32 @@ def main(config_args):
 
     gs.initialize(ip_config=config.ip_config, backend=config.backend,
                   local_rank=config.local_rank)
-
-    infer_data = GSgnnEdgeInferData(config.graph_name,
-                                    config.part_config,
-                                    eval_etypes=config.eval_etype,
-                                    node_feat_field=config.node_feat_name,
-                                    decoder_edge_feat=config.decoder_edge_feat)
+    # language model only encoder does not allow node and edge features
```
Comment (zhjwy9343) [inline python/graphstorm/run/gsgnn_lp/lp_infer_lm.py:41]: ditto

File: python/graphstorm/run/gsgnn_ep/ep_infer_lm.py (hunk 1 above)
Comment (classicsong) [reply python/graphstorm/run/gsgnn_ep/ep_infer_lm.py:49]: Updated

**Review by zhjwy9343 — APPROVED** [review]
LGTM
